src/api/cpp/cvc5.h

```diff
@@ -4372,6 +4486,40 @@ class CVC5_EXPORT Solver
    */
   Statistics getStatistics() const;
 
+  /* .................................................................... */
+  /* Optimization                                                         */
+  /* .................................................................... */
+
+  /**
+   * Run optimization for the objectives
+   * @param objCombination for multi-objective optimization,
+   *  indicate whether it's a BOX optimization,
+   *  a LEXICOGRAPHIC optimization or a PARETO optimization
+   * @return a pair <Result, vector<OptimizationResult>>,
+   *  where Result indicates the overall optimization result,
+   *  for BOX, if any of the objective is UNSAT or UNKNOWN,
+   *    it's UNSAT/UNKNOWN,
+   *  for LEXICOGRAPHIC, if the assertions are UNSAT/UNKNOWN,
+   *    it's UNSAT/UNKNOWN,
+   *  for PARETO, if the assertions are UNSAT
+   *  or the possible solutions are exhausted, it's UNSAT,
+   *    if the assertions are UNKNOWN, it's UNKNOWN;
+   *  vector<OptimizationResult> stores the results of individual objectives
+   */
+  std::pair<Result, std::vector<OptimizationResult>> checkOpt(
+      const std::vector<std::pair<Term, ObjectiveType>>& objectives,
+      ObjectiveCombination objCombination = LEXICOGRAPHIC) const;
+
+  /**
+   * Retrive the next Pareto solution,
+   * valid only after checkOpt is called with PARETO and SAT is returned,
+   * otherwise it's undefined behaviour.
```

Comment:
The use of our API should never result in undefined behavior. We either throw an exception or we perform the action.

Reply:
it's hard to detect whether the situation is ok or not. because you can add more assertions between the checkOpt and nextSolution and it's hard to detect it.

Reply:
even worse, you can change the assertions between the two calls by performing a pop and re-asserting something